### PR TITLE
Add API v2 to our app

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR makes the `h` application aware of "v2" of the API and makes all current endpoints respond to "v2" requests as well as (default) "v1" requests. I had originally thought I'd take care of [documentation first](https://github.com/hypothesis/product-backlog/issues/976) but there's a bit of a chicken-and-egg situation here, so I opted to put in the actual support first.

The diff of this looks large, but it's primarily boilerplate and adding (boilerplate) tests to assure multi-version support. Material things that happen here:

* "v2" is added as a "known version" of the API. Requests that set an `Accept` header to the version 2 media type ("application/vnd.hypothesis.v2+json") are valid.
* The intelligence of the response view decorator for API views has been increased such that it can set the right custom `Hypothesis-Media-Type` response header indicating which version of the API served the request.
* For the `index` endpoint, a new view has been added to support v2. This is merely to illustrate how this would work; both the v1 and v2 views do the same thing other than the v1 view returning v1 service links and the v2 view returning v2 service links (a refactor is in order to DRY these views out a little later, but they're thin so I don't feel terrible with the duplication right now, especially as I'd like to revisit the response for this endpoint for v2).
* For all other API views: they now support both v1 and v2. In other words, at this point, v1 and v2 of our API are effectively identical for the moment.

Documentation is the _very next step_ here, but I wanted to get the thing we're documenting in place first.

Part of https://github.com/hypothesis/product-backlog/issues/940

### CODE REVIEW

File: h/views/api/config.py [hunk 1]
```diff
@@ -78,7 +77,11 @@ def add_api_view(
 
         links.register_link(link, versions, config.registry)
 
-    config.add_view(view=view, **settings)
+    if API_VERSION_DEFAULT in versions:
```
Comment (lyzadanger) [inline h/views/api/config.py]: This fixes a logic bug in which two views could be registered for a single route, both with `accept="application/json"` (which will raise/not be OK), so we need to register only the view that corresponds to the _default_ API version to respond to this accept value.

File: h/views/api/decorators/response.py [hunk 2]
```diff
@@ -4,17 +4,25 @@
...
 
 
 def version_media_type_header(wrapped):
     """View decorator to add response header indicating API version"""
 
     def wrapper(context, request):
         response = wrapped(context, request)
-        # TODO: expand with logic for multiple versions once the app knows
-        # about multiple versions
+        # Assume default version...
```
Comment (lyzadanger) [inline h/views/api/decorators/response.py]: The previous variant of this wrapper was dumb (on purpose at the time) and responded with the media type for the default API version, because there was only one version, and it was the default. Now it will respond more smarter.

File: h/views/api/helpers/links.py [hunk 3]
```diff
@@ -57,7 +57,7 @@ def register_link(link, versions, registry):
         registry.api_links[version].append(link)
 
 
-def format_nested_links(api_links, version, templater):
+def format_nested_links(api_links, templater):
```
Comment (lyzadanger) [inline h/views/api/helpers/links.py]: Derp. There was simply an unused param here.

File: h/views/api/index.py [hunk 4]
```diff
@@ -24,4 +24,20 @@ def index(context, request):
         request.route_url, params=["id", "pubid", "user", "userid", "username"]
     )
 
-    return {"links": link_helpers.format_nested_links(api_links, "v1", templater)}
+    return {"links": link_helpers.format_nested_links(api_links, templater)}
+
+
+@api_config(versions=["v2"], route_name="api.index", link_name="index")
```
Comment (lyzadanger) [inline h/views/api/index.py]: This view function is mostly to illustrate how to respond differently for different versions (and proves that such a thing works.

File: tests/functional/api/groups/test_create.py [hunk 5]
```diff
@@ -19,6 +19,19 @@ def test_it_returns_http_200_with_valid_payload(self, app, token_auth_header):
 
         assert res.status_code == 200
 
+    @pytest.mark.parametrize(
```
Comment (lyzadanger) [inline tests/functional/api/groups/test_create.py]: I did add functests for all endpoints pertaining to their version support. It did help me catch a couple that almost slipped through the cracks, but these functests may be too slow-y down-y to be useful? I dunno.

File: h/views/api/config.py (hunk 1 above)
Comment (lyzadanger) [reply h/views/api/config.py]: This will have a nice side effect of breaking if we try to register more than one view to handle a single version for a single route. We'd want breakage in that case :)

File: h/views/api/decorators/response.py [hunk 6]
```diff
@@ -4,17 +4,25 @@
...
         version_media_type = media_type_for_version(API_VERSION_DEFAULT)
+        # ...Unless we can determine otherwise from the Accept header
+        if request.accept:
+            version_accepts = [t for t in request.accept if t in version_media_types()]
+            if any(version_accepts):
```
Comment (robertknight) [inline h/views/api/decorators/response.py]: Could this be just `if version_accepts` since we only need to test for a non-empty list here.

File: tests/functional/api/groups/test_create.py (hunk 5 above)
Comment (robertknight) [reply tests/functional/api/groups/test_create.py]: I do think testing each and every endpoint separately for multiple version support is redundant and that the costs probably outweigh the benefits. I would be inclined to:

- Have an endpoint to check version support for one or a small number of routes
- Have a test to check the index route output for v1 and v2, since I think this is assembled from the same metadata that is used to implement other aspects of versioning support. If you forgot to add "v2" to one of the API routes, presumably it would not show up in the index route output?

File: tests/h/views/api/index_test.py [hunk 7]
```diff
@@ -30,30 +30,57 @@ def test_it_returns_links_for_the_right_version(
...
+    # declarations.
+    config = Configurator()
+    config.scan("h.views.api.annotations")
+    config.scan("h.views.api.index")
+    # Any route referenced in `h.views.api.annotations` needs to be added here
+    pyramid_config.add_route("api.search", "/dummy/search")
+    pyramid_config.add_route("api.annotations", "/dummy/annotations")
+    pyramid_config.add_route("api.annotation", "/dummy/annotations/:id")
+    pyramid_request.registry.api_links = config.registry.api_links
+    print(config.registry.api_links)
```
Comment (robertknight) [inline tests/h/views/api/index_test.py]: Debug statement left in

File: tests/h/views/api/index_test.py [hunk 8]
```diff
@@ -30,30 +30,57 @@ def test_it_returns_links_for_the_right_version(
...
+
+
+@pytest.fixture
+def pyramid_request(pyramid_config, pyramid_request):
+    # Scan `h.views.api_annotations` for API link metadata specified in @api_config
+    # declarations.
+    config = Configurator()
+    config.scan("h.views.api.annotations")
+    config.scan("h.views.api.index")
+    # Any route referenced in `h.views.api.annotations` needs to be added here
```
Comment (robertknight) [inline tests/h/views/api/index_test.py]: Now that I read this code again it isn't completely obvious to me why this was needed in the first place. I realize that you're just moving some existing code around here though

File: tests/h/views/api/index_test.py (hunk 7 above)
Comment (lyzadanger) [reply tests/h/views/api/index_test.py]: Embarrassing.

Comment (robertknight) [reply tests/h/views/api/index_test.py]: In the client repo we have a lint rule to catch this. I don't know whether a similar thing could easily be turned on for Python?

File: tests/functional/api/groups/test_create.py (hunk 5 above)
Comment (lyzadanger) [reply tests/functional/api/groups/test_create.py]: Yes, let me rethink this. The thing I'm trying to test-protect us for is: inadvertently not having version support for an endpoint. We're protected against having two views trying to provide support for a single version—the opposite problem—because that will raise during config.

Given the way `@api_config` works, all version support can be gleaned from registry entries, I shall endeavor to find a straightforward way to test at this level...

**Comment by lyzadanger** [comment]
@robertknight I've found a relatively straightforward way to test for expected-endpoint support across versions in a contained functional test. Perhaps this could do until we find an approach to integration-like tests in `h`?

File: tests/functional/api/test_versions.py [hunk 9]
```diff
@@ -83,3 +94,28 @@ def test_index_415s_with_invalid_accept_header_value(self, app):
...
+        assert (
+            res.headers["Hypothesis-Media-Type"] == "application/vnd.hypothesis.v1+json"
+        )
+
+    def test_index_adds_v2_response_header(self, app):
+        """
+        Set a v2 Accept header and we should get a version media type
+        response header.
+        """
+        headers = {"Accept": str("application/vnd.hypothesis.v2+json")}
```
Comment (robertknight) [inline tests/functional/api/test_versions.py]: This test looks like it does the same thing as the `test_it_supports_versions` test above. The one above is different because it doesn't supply an `Accept` header.

**Comment by lyzadanger** [comment]
> LGTM. I found a remaining reference to [X-Hypothesis-Media-Type](https://github.com/hypothesis/h/blob/93b02b01a7bfe2b8fca8faf8d6714376c64a422e/docs/_extra/api-reference/hypothesis-v1.yaml#L56) in this branch.

In the docs, looks like. I'll fix in #5592
